File: app/src/main/java/com/alexkn/syntact/app/ViewModelFactory.java
package com.alexkn.syntact.app;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import javax.inject.Inject;
import javax.inject.Provider;

public class ViewModelFactory<VM extends ViewModel> implements ViewModelProvider.Factory {

    private Provider<VM> viewModel;

    @Inject
    public ViewModelFactory(Provider<VM> viewModel) {

        this.viewModel = viewModel;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {

        return (T) viewModel.get();
    }
}

File: app/src/main/java/dev/alexknittel/syntact/presentation/common/ViewModelFactory.kt
package dev.alexknittel.syntact.presentation.common

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

import javax.inject.Inject
import javax.inject.Provider

class ViewModelFactory<VM : ViewModel> @Inject
constructor(private val viewModel: Provider<VM>) : ViewModelProvider.Factory {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T = viewModel.get() as T
}
